File: nodes/node.go
// A capability node written in Go (a compiled, entirely different runtime),
// serving the SAME language-neutral capability contract as the JS node.
// Proves the URI-process "server" is language-independent: same wire protocol,
// same typed contract, different implementation.
package main

import (
	"encoding/json"
	"net/http"
	"os"
	"runtime"
)

type cap struct {
	URI     string      `json:"uri"`
	Effect  string      `json:"effect"`
	Input   interface{} `json:"input"`
	Output  interface{} `json:"output"`
	Backend string      `json:"backend"`
}

func caps() []cap {
	return []cap{{
		URI:    "sys://host/os/query/info",
		Effect: "query",
		Input:  map[string]interface{}{"type": "object", "properties": map[string]interface{}{}},
		Output: map[string]interface{}{
			"type":     "object",
			"required": []string{"ok", "os", "arch", "lang"},
		},
		Backend: "go:" + runtime.Version(),
	}}
}

func writeJSON(w http.ResponseWriter, code int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	json.NewEncoder(w).Encode(v)
}

func main() {
	port := os.Getenv("PORT")
	if port == "" {
		port = "8862"
	}
	mux := http.NewServeMux()
	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, map[string]interface{}{"ok": true, "lang": "go", "capabilities": 1})
	})
	mux.HandleFunc("/capabilities", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, map[string]interface{}{"lang": "go", "capabilities": caps()})
	})
	mux.HandleFunc("/dispatch", func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			URI     string                 `json:"uri"`
			Payload map[string]interface{} `json:"payload"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeJSON(w, 400, map[string]interface{}{"ok": false,
				"error": map[string]string{"category": "INVALID_ARGUMENT", "message": "bad json"}})
			return
		}
		if req.URI != "sys://host/os/query/info" {
			writeJSON(w, 404, map[string]interface{}{"ok": false,
				"error": map[string]string{"category": "NOT_FOUND", "message": "no such capability"}})
			return
		}
		host, _ := os.Hostname()
		writeJSON(w, 200, map[string]interface{}{"ok": true, "result": map[string]interface{}{
			"ok": true, "os": runtime.GOOS, "arch": runtime.GOARCH,
			"hostname": host, "lang": "go",
		}})
	})
	println("capability node (go) on :" + port)
	http.ListenAndServe("0.0.0.0:"+port, mux)
}

File: nodes/node.js
#!/usr/bin/env node
// A capability node (the "server" that executes URI processes) written in
// JavaScript. It speaks the language-neutral capability wire protocol, so a
// client in ANY language drives it with the same typed contract:
//
//   GET  /health        -> { ok, lang, capabilities }
//   GET  /capabilities  -> advertises what contracts it can satisfy (+ backend)
//   POST /dispatch {uri, payload} -> { ok, result } | { ok:false, error }
//
// Implements the SAME contract as the Go node: sys://host/os/query/info
"use strict";
const http = require("http");
const os = require("os");

const CAPS = [{
  uri: "sys://host/os/query/info",
  effect: "query",
  input: { type: "object", properties: {} },
  output: {
    type: "object", required: ["ok", "os", "arch", "lang"],
    properties: {
      ok: { type: "boolean" }, os: { type: "string" },
      arch: { type: "string" }, hostname: { type: "string" }, lang: { type: "string" },
    },
  },
  backend: "node:" + process.version,          // negotiation: how I satisfy it
}];

const HANDLERS = {
  "sys://host/os/query/info": () => ({
    ok: true, os: os.platform(), arch: os.arch(),
    hostname: os.hostname(), lang: "javascript",
  }),
};

function send(res, code, obj) {
  const body = JSON.stringify(obj);
  res.writeHead(code, { "Content-Type": "application/json" });
  res.end(body);
}

http.createServer((req, res) => {
  if (req.method === "GET" && req.url === "/health")
    return send(res, 200, { ok: true, lang: "javascript", capabilities: CAPS.length });
  if (req.method === "GET" && req.url === "/capabilities")
    return send(res, 200, { lang: "javascript", capabilities: CAPS });
  if (req.method === "POST" && req.url === "/dispatch") {
    let data = "";
    req.on("data", (c) => (data += c));
    req.on("end", () => {
      let uri, payload;
      try { ({ uri, payload } = JSON.parse(data || "{}")); }
      catch { return send(res, 400, { ok: false, error: { category: "INVALID_ARGUMENT", message: "bad json" } }); }
      const h = HANDLERS[uri];
      if (!h) return send(res, 404, { ok: false, error: { category: "NOT_FOUND", message: "no such capability" } });
      try { send(res, 200, { ok: true, result: h(payload || {}) }); }
      catch (e) { send(res, 500, { ok: false, error: { category: "INTERNAL", message: String(e) } }); }
    });
    return;
  }
  send(res, 404, { ok: false });
}).listen(process.env.PORT || 8861, "0.0.0.0", () =>
  console.log(`capability node (javascript) on :${process.env.PORT || 8861}`));
